PhysicsTools/NanoAOD/plugins/LeptonTimeLifeInfoTableProducer.cc

```diff
@@ -0,0 +1,334 @@
+/**
+  \class    LeptonTimeLifeInfoTableProducer
+  \brief    Produces FlatTable with lepton life-time information
+
+  \author   Michal Bluj, NCBJ, Warsaw
+*/
+
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/allowedValues.h"
+
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+#include "DataFormats/PatCandidates/interface/Electron.h"
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/PatCandidates/interface/Tau.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+
+#include "MagneticField/Engine/interface/MagneticField.h"
+#include "RecoVertex/VertexPrimitives/interface/TransientVertex.h"
+#include "RecoVertex/VertexTools/interface/VertexDistance3D.h"
+#include "RecoVertex/VertexPrimitives/interface/ConvertToFromReco.h"
+#include "RecoVertex/KalmanVertexFit/interface/KalmanVertexFitter.h"
+#include "TrackingTools/TransientTrack/interface/TransientTrackBuilder.h"
+#include "TrackingTools/Records/interface/TransientTrackRecord.h"
+#include "TrackingTools/GeomPropagators/interface/AnalyticalTrajectoryExtrapolatorToLine.h"
+#include "TrackingTools/GeomPropagators/interface/AnalyticalImpactPointExtrapolator.h"
+#include "CommonTools/Utils/interface/StringCutObjectSelector.h"
+#include "CommonTools/Utils/interface/StringObjectFunction.h"
+
+#include "Utilities/General/interface/ClassName.h"
+#include "PhysicsTools/NanoAOD/interface/DumpedVariable.h"
+#include "DataFormats/VertexReco/interface/TrackTimeLifeInfo.h"
+
+#include <cstring>
+
+template <typename T>
+class LeptonTimeLifeInfoTableProducer : public edm::stream::EDProducer<> {
```

Comment:
I think I did suggest to split the `LeptonTimeLifeInfo` producer and the table producer ; is there anything fundamental preventing this ?

Reply:
I thought I convinced you during the meeting that it is better to produce directly the timeLifeInfo table rather than to produce collections of `LeptonTimeLifeInfo` put them to the Event and then get them back to produce the table. I think that the "split" approach would be justified only when there were other users of the `LeptonTimeLifeInfo` collection(s) than table producer or if there is a strong argument against custom table producers. I am not convinced that any of those is the case. But, I can miss something.
Anyway, it will not be a big deal to move to the "split" approach (which actually was used at initial stage of this development, but with pat leptons as "mediators"). If we decide to go this direction where the LeptonTimeLifeInfo producer should be placed, NanoAOD, PatAlgos other place?

Reply:
Do you agree with my comment or you maintain your opinion about the "splitting"?

Reply:
I understand that this could be complicated due to the track. Maybe keep a full object in `LeptonTimeLifeInfo` instead of a pointer, and make it non persistent

Reply:
After more thinking: I am not sure if it is possible to ensure in the "split" approach that `LeptonTimeLifeInfo` items are sorted in the same way as corresponding leptons in their collections, This prevents extending simply lepton tables by adding time-life info columns... In other words this approach forces adding new time-life info tables which then can be linked with lepton ones by indexes. This is doable, but it is not what was originally intended. Is it what is requested?

Reply:
you may use valuemap ?

Reply:
@vlimant Could you please give more details about proposed solution with valuemap? Values maps for simple numerical types can be easily passed to extend tables of their "parent" collections with `externalVariables`, but there is not a way to do this with variables to be extracted from complex types/data-formats (at least I do not see how to). It can be probably implemented, but it is not straightforward as type of the complex data-format has to be passed and then phrased by a table filler.
Solution with producing a collection of time-life infos and then producing table on top of it with a simple-table-producer does not guarantee that there will be correct matching between infos and their parent leptons...
Maybe you propose to produce valuemap of indexes (`int`s) which can be used for matching between leptons and items of time-life info collection? Indexes then can be stored in lepton tables and can point to items in separate time-life info tables. This should work, however, it is not very elegant.